### PR TITLE
Refactor Conformance Test Generation for better test logic reuse and shallower test tree.

### PR DESCRIPTION
**Pull request by jpschorr**
### Refactor test logic
- Do not generate top-level module. Instead it is written and checked in.
- Provide `pass_*` and `fail_*` test methods in top-level that all tests delegate to.

### Refactor directory/file layout
- Encapsulated Writer
- Add 'banner' to generated files telling user 'do not edit'
- Allow only generating upto a certain tree depth in the filesystem (shortens and broadens the tree of files)

---

new: 
``` shell
❯ tree tests
tests
├── mod.rs
└── partiql_tests
    ├── fail
    │   ├── mod.rs
    │   ├── static_analysis
    │   │   ├── mod.rs
    │   │   └── primitives.rs
    │   └── syntax
    │       ├── mod.rs
    │       ├── primitives.rs
    │       └── query.rs
    ├── mod.rs
    └── success
        ├── mod.rs
        └── syntax
            ├── mod.rs
            ├── primitives.rs
            └── query.rs

6 directories, 12 files
```
old (abridged):
```shell
❯ tree tests_v1
tests_v1
├── mod.rs
└── partiql_tests
    ├── mod.rs
    └── partiql_tests_data
        ├── fail
        │   ├── mod.rs
        │   ├── static_analysis
             <snip>
        │   └── syntax
        │       ├── mod.rs
        │       ├── primitives
        │       │   ├── call.rs
        │       │   ├── case.rs
        │       │   ├── cast.rs
        │       │   ├── container_constructors.rs
        │       │   ├── date_constructor.rs
        │       │   ├── expressions.rs
        │       │   ├── mod.rs
        │       │   ├── operators
        │       │   │   ├── at_operator.rs
        │       │   │   ├── between_operator.rs
        │       │   │   ├── like_operator.rs
        │       │   │   └── mod.rs
        │       │   ├── path_expression.rs
        │       │   └── time_constructor.rs
        │       └── query
        │           ├── mod.rs
        │           ├── pivot.rs
        │           └── select
        │               ├── joins.rs
        │               ├── limit_offset.rs
        │               ├── mod.rs
        │               ├── order_by.rs
        │               └── select.rs
    
        <snip>

17 directories, 73 files
```
new `primitives.rs` (abridged): 
```rust
<snip>

mod cast {
    #[test]
    fn r#cast_expected_left_paren() {
        crate::fail_syntax(r#"CAST 5 as integer"#);
    }

    #[test]
    fn r#cast_given_keyword_as_type_arg() {
        crate::fail_syntax(r#"CAST(5 AS SELECT)"#);
    }
}

<snip>
```

old `cast.rs`: 
```rust
#[test]
fn r#cast_expected_left_paren() {
    let statement = r#"CAST 5 as integer"#;
    let res = partiql_parser::Parser::default().parse(statement);
    assert!(
        res.is_err(),
        "For `{}`, expected `Err(_)`, but was `{:#?}`",
        statement,
        res
    );
}

#[test]
fn r#cast_given_keyword_as_type_arg() {
    let statement = r#"CAST(5 AS SELECT)"#;
    let res = partiql_parser::Parser::default().parse(statement);
    assert!(
        res.is_err(),
        "For `{}`, expected `Err(_)`, but was `{:#?}`",
        statement,
        res
    );
}
```

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-conformance-test-generator/src/writer.rs [hunk 1]
```diff
@@ -1,64 +1,131 @@
...
+        &self,
+        path: impl AsRef<Path>,
+        module: TestModule,
+    ) -> miette::Result<()> {
+        let name = path.as_ref().file_name().unwrap().to_str().unwrap();
+        let mut module = collapse_module(name, module)?;
 
-fn write_top_level(path: impl AsRef<Path>, sub_tests_dir: &str) -> miette::Result<()> {
-    std::fs::create_dir_all(&path).into_diagnostic()?;
+        let mut path = PathBuf::from(path.as_ref()).with_extension("rs");
```
Comment (alancai98) [inline partiql-conformance-test-generator/src/writer.rs:110]: (warning from GitHub): `path` doesn't need to be mutable

File: partiql-conformance-test-generator/src/writer.rs [hunk 2]
```diff
@@ -1,64 +1,131 @@
...
+    }
 
-                write_module(child_path, module)?
+    fn write_module(
+        &self,
+        path: impl AsRef<Path>,
+        module: TestModule,
+        depth: u8,
+    ) -> miette::Result<()> {
+        println!("####{}::{:?}", depth, &path.as_ref());
```
Comment (alancai98) [inline partiql-conformance-test-generator/src/writer.rs:74]: `println!` could be removed?

File: partiql-conformance-test-generator/src/writer.rs [hunk 3]
```diff
@@ -80,10 +147,10 @@ fn write_dir_mod<'a>(
 }
 
 fn write_scope(path: impl AsRef<Path>, scope: &Scope) -> miette::Result<()> {
-    let contents = scope.to_string();
+    println!("write_scope-{:?}", path.as_ref());
```
Comment (alancai98) [inline partiql-conformance-test-generator/src/writer.rs:150]: `println!` could be removed?

File: partiql-conformance-test-generator/src/reader.rs [hunk 4]
```diff
@@ -13,19 +8,17 @@ use std::ffi::OsStr;
 use std::fs;
 use std::fs::DirEntry;
 
-use std::ops::Add;
+use crate::schema::spec::{
+    Assertion, Assertions, Namespace, Namespaces, TestCase, TestCases, TestDocument,
+};
+use crate::schema::structure::{TestDir, TestEntry, TestFile, TestRoot};
 use std::path::{Path, PathBuf};
```
Comment (alancai98) [inline partiql-conformance-test-generator/src/reader.rs:15]: (warning from GitHub/rust compiler): `PathBuf` is unused

**Review by alancai98 — APPROVED** [review]
Looks good. The generated test files + test directory structure looks a lot more polished now. Left a few minor nits.

File: partiql-conformance-test-generator/src/writer.rs (hunk 2 above)
Comment (jpschorr) [reply partiql-conformance-test-generator/src/writer.rs:74]: whoops

File: partiql-conformance-test-generator/src/writer.rs (hunk 3 above)
Comment (jpschorr) [reply partiql-conformance-test-generator/src/writer.rs:150]: whoops

**Pull request merged**